### PR TITLE
More aggresive version of BatchSpanProcessor

### LINKED ISSUE #729: Simplify the BatchSpansProcessor
I believe this class can be re-implemented to use a simple Queue and a background thread to process it.

### LINKED ISSUE #1560: Replace usages of java.util.Timer with a ScheduledExecutor in the BatchSpanProcessor
The Timer has known issues with clock changes. We should use ScheduledExecutors as they are the more modern construct for doing timed executions.

### LINKED ISSUE #1566: Leverage BatchSpanProcessor's span dropping functionality
The [recent change to make the export methods async](https://github.com/open-telemetry/opentelemetry-java/pull/1428) appears to have introduced a situation where the `BatchSpanProcessor` can lose telemetry in the case where its `spansList` isn't drained fast enough.

While the `BatchSpanProcessor` is built to cater for spans being dropped, and therefore its usage must assume this as a possibility, we can at least ensure that `exportBatches` is replaced by a function to take just one batch (`drainBatch`?) so that the existing `BatchSpanProcessor` reporting of dropped telemetry is leveraged.

### PR DESCRIPTION
**Pull request by iNikem**
Alternative to #1571

Fixes #729
Fixes #1566
Fixes #1560

### CODE REVIEW

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java
Comment (anuraaga) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java:127]: `CompletableFuture.join` doesn't throw interrupted exception so I think it's ok to squash that here too (don't know if it's best practice for interruptions but may as well copy the JDK)

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java
Comment (anuraaga) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java:114]: I think we can leave out this method, joining without timeout is rarely a good idea.

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (anuraaga) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:199]: Is there a reason to wait for a constant here instead of until the next export time?

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (anuraaga) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:202]: It's probably a nice win to add a `drainTo` here to get as many elements as possible. In the case where there are constantly spans, we'll be able to handle them with less queue lock acquisitions and tighter loop. `drainTo` has this very tight loop inside the lock which JVM should deal well with.

```java
while (i < n) {
                    @SuppressWarnings("unchecked")
                    E e = (E) items[take];
                    c.add(e);
                    items[take] = null;
                    if (++take == items.length) take = 0;
                    i++;
                }
```

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (anuraaga) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:219]: Similar to above, it might be better if we can structure this to take advantage of drainTo. I guess you can consecutively drainTo and subtract the number drained from spansToFlush instead of individual poll without a huge change in structure.

**Comment by anuraaga** [comment]
By the way I added this benchmark https://github.com/open-telemetry/opentelemetry-java/pull/1571/files#diff-85f8a2fe8ea17909437002d022e86fab - no clue if it's a very good benchmark but maybe it can be useful :)

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (iNikem) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:259]: I am not sure about that. We have to prevent concurrent calls to export and we have to have timeouts. IMO timed joins is the easiest way to achieve this. At least until benchmarking shows that this is too expensive.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java
Comment (iNikem) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java:127]: You mean catch that inside the method?

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java
Comment (iNikem) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java:114]: I use it `forceFlush`. There is no good option how long should I wait there. And if you say that `forceFlush` should be async, I will just change my argument to `shutdown` :)

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (iNikem) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:199]: Hm... I will think about that. My first concern is the introduction of `System.nanoTime` on essentially every `poll`.

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (iNikem) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:219]: I will try to benchmark this class today

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (iNikem) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:199]: Poll certainly cannot block for `scheduleDelayMillis` if it is already almost time to export, but we have partial batch.

**Comment by jkwatson** [comment]
> By the way I added this benchmark https://github.com/open-telemetry/opentelemetry-java/pull/1571/files#diff-85f8a2fe8ea17909437002d022e86fab - no clue if it's a very good benchmark but maybe it can be useful :)

What do you think about putting in that benchmark as a stand-alone PR, so we can test various solutions on the same footing?

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (iNikem) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:202]: A, main reason I cannot use `drain` here is that I immediately convert to `SpanData`. And I don't want to drain to one collection, then iterate and convert element by element into another collection.

**Comment by iNikem** [comment]
Pushed some polish and my take on @anuraaga 's benchmark. I was more interested in the dropped spans count. Unfortunately could not figure out how to properly measure that, but current benchmark gives some indication. We would greatly benefit from utility to emit new spans with given rate...

Anyway, how should we proceed and choose between different proposal for BSP? Let me remind you, that java instrumentation fails tests with current otel-java snapshot, because of BSP.

File: build.gradle
Comment (jkwatson) [inline build.gradle:73]: what is the use-case for an unused nested class that we wouldn't want to warn about it?

Comment (iNikem) [reply build.gradle:73]: Damn... My benchmark class has a `ThreadState` inner class that error prone disliked. I will take a look if I can remove this.

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (iNikem) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:228]: See https://github.com/open-telemetry/opentelemetry-java/pull/1579#discussion_r475713805

File: build.gradle
Comment (iNikem) [reply build.gradle:73]: Removed this line

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (jkwatson) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:213]: this is very similar to what I did when I took a swing at this class 6 months ago, so obviously I think it's a great approach. ;)

**Comment by iNikem** [comment]
After discussion with @anuraaga we have decided to move forward with this PR. In addition I will file a new issue for making `SpanProcessor.forceFlush` and `SpanProcessor.shutdown` methods non-blocking.

**Comment by iNikem** [comment]
Just in case I remind that this and #1585 are required to let instrumentation upgrade to the latest otel-java version. And we need that to move forward

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java
Comment (anuraaga) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java:122]: If you don't mind copying in #1583 that'd be cool, I think the synchronous case optimization and less branching (same lines of code to just all `fail()` twice) is worth it. Then no worries closing that PR in favor of this one

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (anuraaga) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:210]: AINTD, we still have the issue that a flush can be delayed by a long 100ms. Do you mind filing an issue for it? Flush is mostly used on FaaS so they pay $$ for the time they use, we should put in the effort to save that even if it adds some complexity.

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (anuraaga) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:214]: I think it's only this line that can produce a `InterruptedException`, can you narrow down the scope of the try / catch to make that more obvious?

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (anuraaga) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:234]: As we discussed in chat, I still have reservations about the lock contention between `.poll` and `.offer`. It'd be nice if we can file an issue about performance benchmarking / optimization of BSP. If I understood correctly, we're ok with adding complexity if we start with numbers and show the numbers with it, which makes sense to me.

**Review by anuraaga — APPROVED** [review]
I'm good with this approach. I consider it the simplest, but it's likely not the most optimal. We should iterate on it with some clear numeric targets, and I think it's reasonable for us to have extension artifacts such as an `opentelemetry-exporter-rxjava` that uses a nice async library for handling it even better. Of course, such extensions are blocked on #1584 which is why that's an important one which I'll rebase #1571 to address.

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (iNikem) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:210]: I will do that after this PR is merged. It will be easier to refer to the code concerned.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java
Comment (iNikem) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java:122]: Will do.

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (iNikem) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:234]: #1587 1587

File: sdk/all/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorDroppedSpansBenchmark.java
Comment (jkwatson) [inline sdk/all/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorDroppedSpansBenchmark.java:101]: 😂

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java
Comment (jkwatson) [inline sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java:131]: same comment as on @anuraaga 's PR: can we just do `if (isDone()) return this;`  instead of duplicating the code?

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (jkwatson) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:158]: what would you think about having only one counter called "processedSpans", with a boolean label for dropped or not?

**Comment by jkwatson** [comment]
Generally looking good. Just 2 small questions/changes requested.

File: sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java
Comment (iNikem) [reply sdk/common/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java:131]: Rebased and took all changes from Anuraag PR

**Review by anuraaga — APPROVED** [review]
🙇

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (jkwatson) [inline sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:152]: this description isn't quite right any more. Can be fixed in a follow-on PR, to get you unblocked.

File: sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java
Comment (huntc) [reply sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java:267]: Join shouldn’t have a timeout. There’s nothing you can really do if you can’t join. This is why CompleteableFuture and even back to pthreads has no join.

If an exporter doesn’t return then there is nothing that can be done here other than fail it’s result. Exporters can then act on a result failure and halt their activity.

**Review by huntc — COMMENTED** [review]
It’s really disappointing that we finally reinstated blocking. It is unnecessary to block.

**Comment by iNikem** [comment]
@huntc Are you talking about blocking when calling `exporter.export` method? Can you explain to me why is this bad? I mean that method is called from the dedicated worker thread and we have to ensure that `export` method is not called concurrently. I don't see why blocking is bad in this situation?

**Comment by huntc** [comment]
> @huntc Are you talking about blocking when calling `exporter.export` method? Can you explain to me why is this bad? I mean that method is called from the dedicated worker thread and we have to ensure that `export` method is not called concurrently. I don't see why blocking is bad in this situation?

I was hoping for an executor that can be shared across many components. This is no longer possible without hogging a thread from an executor used in other places. Reusing executors saves on resources.

**Comment by iNikem** [comment]
You mean not using a dedicated BSP-specific worker thread, but a shared executor for all internal concurrent needs? This is a good idea :) And you are absolutely correct, if this idea is ever to be implemented, we will need to change this implementation to non-blocking one.

**Pull request merged**
